### PR TITLE
Two ways to calculate NK

### PR DESCRIPTION
**Pull request by andrewram4287**

### CODE REVIEW

**Comment by aarmey** [comment]
@andrewram4287 can you write a test that calls this function with inputs of the right shapes? You can fill them with dummy numbers—like a matrix of ones. This will be helpful for ensuring it is setup correctly.

File: gmm/tests/test_GMM.py [hunk 1]
```diff
@@ -135,3 +135,15 @@ def test_cov_fit():
     assert math.isclose(cov[1][0], covR[1][0], abs_tol=0.2)
     assert math.isclose(cov[0][1], covR[0][1], abs_tol=0.2)
     assert math.isclose(cov[1][1], covR[1][1], abs_tol=0.3)
+    
+def test_loglikelihood():
```
Comment (aarmey) [inline gmm/tests/test_GMM.py:139]: This function should directly call `comparingGMMjax_NK()` so you are only testing that code. Like I mentioned, you can make dummy arguments from scratch. The current code cannot be correct, because `nk` should be differently sized depending on whether it is a factor matrix or vector describing the constant values.

**Review by aarmey — CHANGES_REQUESTED** [review]

File: gmm/tensor.py [hunk 2]
```diff
@@ -95,6 +95,29 @@ def comparingGMMjax(X, nk, meanFact: list, tPrecision):
     loglik = jnp.sum(jsp.logsumexp(log_prob + log_det[jnp.newaxis, :, :, :, :] + nkl[jnp.newaxis, :, jnp.newaxis, jnp.newaxis, jnp.newaxis], axis=1))
     return loglik
 
+def comparingGMMjax_NK(X, nk, meanFact: list, tPrecision):
+    """Obtains the GMM means, convariances and NK values along with zflowDF mean marker values
+    to determine the max log-likelihood"""
+    assert nk.ndim == 1
```
Comment (aarmey) [inline gmm/tensor.py:101]: Isn't `nk` now supposed to be a factor matrix? In that case, it cannot be a vector.

File: gmm/tensor.py [hunk 3]
```diff
@@ -95,6 +95,29 @@ def comparingGMMjax(X, nk, meanFact: list, tPrecision):
...
+    n_markers = tPrecision.shape[1]
+    nkl = jnp.log(nk / jnp.sum(nk))
```
Comment (aarmey) [inline gmm/tensor.py:103]: I had noted about this needing to have an `axis` argument, because `nk` is now a matrix.

File: gmm/tensor.py [hunk 4]
```diff
@@ -179,9 +206,13 @@ def sample_GMM(weights_, means_, cholCovs, n_samples):
     n_samples_comp = np.random.multinomial(n_samples, weights_)
 
     X = np.vstack(
-        [np.random.multivariate_normal(mean, cholCov @ cholCov.T, int(sample))
+        [
```
Comment (aarmey) [inline gmm/tensor.py:209]: Did this change or is this just linting?

Comment (andrewram4287) [reply gmm/tensor.py:209]: Just linting I believe

File: gmm/tensor.py (hunk 2 above)
Comment (andrewram4287) [reply gmm/tensor.py:101]: Ok I will change that!

File: gmm/tests/test_GMM.py (hunk 1 above)
Comment (andrewram4287) [reply gmm/tests/test_GMM.py]: Ok I think i was able to alter it correctly based on the tMeans tensor.

File: gmm/tensor.py [hunk 5]
```diff
@@ -96,10 +96,40 @@ def comparingGMMjax(X, nk, meanFact: list, tPrecision):
...
+    assert nkFact.ndim == 2
+    nk = nkFact @ meanFact[2].T
+    assert nk.ndim == 2
+    nkl = jnp.log(nk / jnp.sum(nk, axis=0, keepdims=True))
+    print(len(meanFact))
+    print(meanFact[0].shape)
+    print(meanFact[1].shape)
+    print(meanFact[2].shape)
+    print(tPrecision.shape)
+    mp = jnp.einsum("iz,jz,kz,ijok->iok", *meanFact, tPrecision)
```
Comment (aarmey) [inline gmm/tensor.py:112]: `meanFact` still has five dimensions, so this reduced `einsum` is failing.

File: gmm/tests/test_GMM.py [hunk 6]
```diff
@@ -135,3 +135,15 @@ def test_cov_fit():
     assert math.isclose(cov[1][0], covR[1][0], abs_tol=0.2)
     assert math.isclose(cov[0][1], covR[0][1], abs_tol=0.2)
     assert math.isclose(cov[1][1], covR[1][1], abs_tol=0.3)
+    
+def test_loglikelihood_NK():
+    """Testing to see if loglilihood is a number"""
+    PAShape = (6, dataPA_import.shape[0], dataPA_import.shape[2],dataPA_import.shape[3], dataPA_import.shape[4])
+    x0 = vector_guess(PAShape, rank=3)
+    _, meanFact,covFac = vector_to_cp_pt(x0, rank=3, shape=PAShape)
```
Comment (aarmey) [inline gmm/tests/test_GMM.py:143]: You should build `meanFact` and `precBuild` manually, like I suggested. `vector_to_cp_pt` does not provide the right shapes.

**Review by aarmey — CHANGES_REQUESTED** [review]

File: gmm/tensor.py (hunk 5 above)
Comment (andrewram4287) [reply gmm/tensor.py:112]: I was able to change the shape of meanFact and tPrecision! But now the einsum seems to be failing

**Review by aarmey — APPROVED** [review]
I think this was an issue with the test. I have a fix here that is passing. Please look everything over and, if it all makes sense, you can merge.
